File: StripMate/Services/Giphy/GiphyService.swift
import Foundation

// MARK: - GIPHY Sticker Model

public struct GiphySticker: Identifiable, Sendable {
    public let id: String
    public let previewUrl: String  // fixed_height_small — picker grid
    public let originalUrl: String // original — saved to Firestore
    public let title: String
}

// MARK: - GIPHY REST API Service

public actor GiphyService {
    public static let shared = GiphyService()

    // TODO: Replace with your GIPHY API key from https://developers.giphy.com/
    private let apiKey = "gFftw3FuFBXmeWPC9x7N4fbXQFcYvnGL"
    private let baseUrl = "https://api.giphy.com/v1/stickers"
    private let session = URLSession.shared

    /// Search GIPHY stickers with a query.
    public func searchStickers(query: String, limit: Int = 30, offset: Int = 0) async throws -> [GiphySticker] {
        guard !query.isEmpty else { return try await trendingStickers(limit: limit) }
        let encoded = query.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? query
        let url = URL(string: "\(baseUrl)/search?api_key=\(apiKey)&q=\(encoded)&limit=\(limit)&offset=\(offset)&rating=pg")!
        return try await fetch(url: url)
    }

    /// Fetch trending GIPHY stickers.
    public func trendingStickers(limit: Int = 30, offset: Int = 0) async throws -> [GiphySticker] {
        let url = URL(string: "\(baseUrl)/trending?api_key=\(apiKey)&limit=\(limit)&offset=\(offset)&rating=pg")!
        return try await fetch(url: url)
    }

    // MARK: - Private

    private func fetch(url: URL) async throws -> [GiphySticker] {
        let (data, response) = try await session.data(from: url)
        guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
            throw GiphyError.invalidResponse
        }

        let json = try JSONSerialization.jsonObject(with: data) as? [String: Any]
        guard let dataArray = json?["data"] as? [[String: Any]] else {
            throw GiphyError.parseError
        }

        return dataArray.compactMap { item -> GiphySticker? in
            guard let id = item["id"] as? String,
                  let images = item["images"] as? [String: Any] else { return nil }

            // Preview: fixed_height_small for fast loading in grid
            let preview = images["fixed_height_small"] as? [String: Any]
            let previewUrl = preview?["url"] as? String ?? ""

            // Original: full quality for storing
            let original = images["original"] as? [String: Any]
            let originalUrl = original?["url"] as? String ?? ""

            guard !previewUrl.isEmpty, !originalUrl.isEmpty else { return nil }

            let title = item["title"] as? String ?? ""

            return GiphySticker(id: id, previewUrl: previewUrl, originalUrl: originalUrl, title: title)
        }
    }
}

// MARK: - Errors

public enum GiphyError: LocalizedError {
    case invalidResponse
    case parseError

    public var errorDescription: String? {
        switch self {
        case .invalidResponse: return "GIPHY yanıt vermedi."
        case .parseError: return "GIPHY verileri okunamadı."
        }
    }
}

File: android/app/src/main/java/com/celalbasaran/stripmate/service/giphy/GiphyService.kt
package com.celalbasaran.stripmate.service.giphy

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder
import javax.inject.Inject
import javax.inject.Singleton

/**
 * GIPHY sticker model for search/trending results.
 */
data class GiphySticker(
    val id: String,
    val previewUrl: String,   // fixed_height_small — picker grid
    val originalUrl: String,  // original — saved to Firestore
    val title: String
)

/**
 * REST API client for GIPHY sticker search & trending.
 * Uses the same API key as the iOS app (Secrets.plist).
 */
@Singleton
class GiphyService @Inject constructor() {

    private val apiKey = "gFftw3FuFBXmeWPC9x7N4fbXQFcYvnGL"
    private val baseUrl = "https://api.giphy.com/v1/stickers"

    /**
     * Search GIPHY stickers with a query. Falls back to trending if query is blank.
     */
    suspend fun searchStickers(
        query: String,
        limit: Int = 30,
        offset: Int = 0
    ): List<GiphySticker> {
        if (query.isBlank()) return trendingStickers(limit, offset)
        val encoded = URLEncoder.encode(query, "UTF-8")
        val url = "$baseUrl/search?api_key=$apiKey&q=$encoded&limit=$limit&offset=$offset&rating=pg"
        return fetch(url)
    }

    /**
     * Fetch trending GIPHY stickers.
     */
    suspend fun trendingStickers(
        limit: Int = 30,
        offset: Int = 0
    ): List<GiphySticker> {
        val url = "$baseUrl/trending?api_key=$apiKey&limit=$limit&offset=$offset&rating=pg"
        return fetch(url)
    }

    private suspend fun fetch(urlString: String): List<GiphySticker> = withContext(Dispatchers.IO) {
        val connection = URL(urlString).openConnection() as HttpURLConnection
        try {
            connection.requestMethod = "GET"
            connection.connectTimeout = 10_000
            connection.readTimeout = 10_000

            if (connection.responseCode != 200) return@withContext emptyList()

            val responseText = connection.inputStream.bufferedReader().readText()
            val json = JSONObject(responseText)
            val dataArray = json.optJSONArray("data") ?: return@withContext emptyList()

            val stickers = mutableListOf<GiphySticker>()
            for (i in 0 until dataArray.length()) {
                val item = dataArray.optJSONObject(i) ?: continue
                val id = item.optString("id", "")
                val images = item.optJSONObject("images") ?: continue

                val preview = images.optJSONObject("fixed_height_small")
                val previewUrl = preview?.optString("url", "") ?: ""

                val original = images.optJSONObject("original")
                val originalUrl = original?.optString("url", "") ?: ""

                if (previewUrl.isNotBlank() && originalUrl.isNotBlank()) {
                    stickers.add(
                        GiphySticker(
                            id = id,
                            previewUrl = previewUrl,
                            originalUrl = originalUrl,
                            title = item.optString("title", "")
                        )
                    )
                }
            }
            stickers
        } finally {
            connection.disconnect()
        }
    }

    companion object {
        /** Check if a message text is a GIPHY URL */
        fun isGiphyUrl(text: String): Boolean {
            val trimmed = text.trim()
            return trimmed.contains("giphy.com/") || trimmed.contains("media.giphy.com/")
        }
    }
}
